Support custom autograd functions in C++

malvika2147 opened the pull request:
### All the PRs in this stack have been moved to #23572, abandoning this PR

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23184 API for AutogradContext
* **#23020 Support custom autograd functions in C++**

Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. For example: 

```
class MyFunction : public Function<MyFunction> {
  static variable_list forward(AutogradContext *ctx, int n, Variable var);
  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

This is a first draft of what the API will look like. 
Planned updates to this stack: 
1. ~~Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`.~~ #23184 
2. ~~Support for non variable inputs to `forward` and to save in context~~
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

Differential Revision: [D16568636](https://our.internmc.facebook.com/intern/diff/D16568636)

Commit:
Support custom autograd functions in C++

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -3,12 +3,19 @@
 namespace torch { namespace autograd {
 
-std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &inputs,
+variable_list _wrap_outputs(const variable_list &input_vars,
```
OK :) This seems like it might be a reasonable change to land ahead of time before all of the custom autograd functions?

ezyang commented:
```
class MyFunction : public CFunction {
```

Don't you need CRTP here? So it will look like:

```
class MyFunction : public CFunction<MyFunction> {
```

Otherwise you won't be able to get the shared `apply` implementation to call the static methods.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -5,11 +5,41 @@
+
+template <class T>
+struct CFunction {
+  static void apply(variable_list&& inputs);
```
You're probably going to want a non rvalue-reference version of this overload, for cases when people don't want to move their variable list into apply (similar to how Function is setup)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -78,7 +85,42 @@ std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &i
+  bool is_executable =  GradMode::is_enabled() && any_variable_requires_grad(input_vars);
+  auto next_edges = collect_next_edges(input_vars);
+
+  std::shared_ptr<CustomFunc<T>> grad_fn = std::make_shared<CustomFunc<T>>();
```
You're going to want to use our custom deleter, `deleteFunction`. Example https://github.com/pytorch/pytorch/pull/22983/files#diff-3379b66ac18dc024127ee0a89616eb7eR669 and the sordid backstory at #6873

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -78,7 +85,42 @@ std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &i
+variable_list CustomFunc<T>::apply(variable_list&& inputs) {
+  auto outputs = T::backward(&ctx, inputs);
 
+  if (outputs.size() != inputs.size()) {
```
This check here differs from the python function implementation (which accepts a longer outputs list as long as the later outputs are undefined.) But I don't know why we did that in the first place, so the stricter check here seems fine. (@apaszke, maybe you know?)

yf225 replied:
I suspect it's because we want to support the following use case:
https://github.com/pytorch/pytorch/blob/425d28c30abcac22d9bdd2d01cc83e3e0470eb11/test/test_autograd.py#L2131-L2138
Here there are some inputs to the forward function that don't need to participate in the autograd graph, but we still want the number of outputs from the custom backward function to match the number of inputs into the custom forward function.

yf225 replied:
Also it would be great to minimize the behavioral difference between Python and C++ API, so that it's easier for people to convert their code to C++ :D

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -78,7 +85,42 @@ std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &i
+    outputs = T::forward(&grad_fn->ctx, input_vars);
+  }
+
+  _wrap_outputs(input_vars, grad_fn.ctx->non_differentiable, grad_fn.ctx->dirty_inputs, outputs, is_executable ? grad_fn : nullptr);
```
I suppose the bit after here is still WIP :)

ezyang commented:
This looks like a good start! And it is pleasingly simple too. Looking forward to seeing the tests.

Commit:
Update on "[WIP] Support custom autograd functions in C++"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23020 Support custom autograd functions in C++**

Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of CFunction:
```
class MyFunction : public CFunction<MyFunction> {
  static variable_list forward(AutogradContext *ctx, variable_list inputs);

  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
`MyFunction::apply(inputs)
`

This is a first draft of what the API will look like. Currently `forward` only accepts `Variable` inputs.
Planned updates to this stack: 
1. Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`. 
2. Support for non variable inputs to `forward` and to save in context
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -86,6 +86,84 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+}
+
+template <typename... Args>
+void extract_vars(variable_list& list, Variable& cur, Args&& ... args) {
```
I'm not sure you wanted a non-const reference here?

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,49 @@ TORCH_API variable_list _wrap_outputs(
+// MyFunction::apply(6, x);
+
+template <class T>
+struct CFunction {
```
naming bikeshedding: since it's called `torch.autograd.Function` in Python, it would be best if we could similarly expose this as `torch::autograd::Function`. Unfortunately, that name is already taken here. I wonder if some renaming isn't in order...

albanD replied:
Not sure about the naming here as well.
In cpp, the current `Function` (and it's subclasses) are nodes in the graph we build.
In python, `Function` can be seen as the definition of an autograd atomic operation.
The problem is that both meanings are quite deeply rooted now no?

But I agree that in cpp, the current `Function` should become `Node` (and `PyNode`, `CppNode`etc) and this class should become `Function`.
Would it be ok for backward compatibility?

malvika2147 replied:
Renaming in https://github.com/pytorch/pytorch/pull/23269

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,49 @@ TORCH_API variable_list _wrap_outputs(
+  static variable_list apply(Args&&... args);
+};
+
+struct AutogradContext {
```
Gonna want some comments here eventually ;)

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,49 @@ TORCH_API variable_list _wrap_outputs(
+  std::unordered_set<at::TensorImpl*> dirty_inputs;
+};
+
+template <class T>
```
Docs here too

ezyang replied:
Pretty please? :)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -86,6 +86,84 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+template<typename... Args>
+variable_list CFunction<T>::apply(Args&&... args) {
+  variable_list input_vars;
+  extract_vars(input_vars, std::forward<Args>(args)...);
```
I think you don't actually want a `std::forward` here; if you have an rvalue reference, you don't want to move it into `extract_vars` (TBF, no move will actually happen because you don't actually do anything with those arguments; it's more about the principle of the thing.)

ezyang replied:
@smessmer maybe you have a better idea?

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,49 @@ TORCH_API variable_list _wrap_outputs(
+};
+
+template <class T>
+struct CustomFunc : public Function {
```
Why not call this `CppFunction`. This is a specification of `Function` for cpp the same way `PyFunction` is a specialization for python?

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,49 @@ TORCH_API variable_list _wrap_outputs(
+  static variable_list apply(Args&&... args);
+};
+
+struct AutogradContext {
```
Is there any way in way to allow the user to store arbitrary data in the ctx between the forward and the backward the same way it can be done in python?
Even if it just subclassing this, it should be in the example.

malvika2147 replied:
Yes, I have added another pr for AutogradContext : https://github.com/pytorch/pytorch/pull/23184

albanD reviewed (commented):
This is going to be super useful, thanks !!

ezyang commented:
Removing WIP tag, I guess.

ezyang commented:
I'll give this another closer review when tests show up ;)

Commit:
Update on "Support custom autograd functions in C++"


Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of CFunction, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. For example: 

```
class MyFunction : public CFunction<MyFunction> {
  static variable_list forward(AutogradContext *ctx, int n, Variable var);
  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

This is a first draft of what the API will look like. 
Planned updates to this stack: 
1. ~~Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`.~~ #23184 
2. ~~Support for non variable inputs to `forward` and to save in context~~
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

Commit:
Update on "Support custom autograd functions in C++"


Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of CFunction, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. For example: 

```
class MyFunction : public CFunction<MyFunction> {
  static variable_list forward(AutogradContext *ctx, int n, Variable var);
  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

This is a first draft of what the API will look like. 
Planned updates to this stack: 
1. ~~Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`.~~ #23184 
2. ~~Support for non variable inputs to `forward` and to save in context~~
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -3,6 +3,21 @@
 
 namespace torch { namespace autograd {
 
+VariableInfo::VariableInfo(const Variable& var)
```
No substantive change here, is that right? Just code motion?

malvika2147 replied:
yes

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+  std::unordered_set<at::TensorImpl*> dirty_inputs;
+};
+
+struct TORCH_API VariableInfo {
```
Also code motion

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+}
+
+template<class T>
+variable_list CppNode<T>::apply(variable_list&& inputs) {
```
I'd recommend commenting here that this logic is the same as in `PyNode::apply` and so if you modify here, you should modify PyNode too. (And vice versa!)

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+    msg += name() + " returned an incorrect number of gradients (expected ";
+    msg += std::to_string(num_forward_inputs) + ", got " ;
+    msg += std::to_string(num_outputs) + ")";
+    throw std::runtime_error(msg);
```
I know you lifted this straight from the old code (so, no action needed in this PR), but we can modernize the error reporting here, using the `TORCH_CHECK` macro.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+        std::string msg("function ");
+        msg += name() + " returned a gradient different that is defined at position ";
+        msg += std::to_string(i + 1) + ", but the corresponding forward input was not a Variable";
+        throw std::runtime_error(msg);
```
Same deal with modernized error reporting here.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+
+  const size_t num_inputs = sizeof...(Args);
+  input_vars.reserve(num_inputs);
+  node->is_variable_input.reserve(num_inputs);
```
We should put a TODO here to add tracing at some future point in time.

ezyang reviewed (approved):
Lovely!

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+  std::vector<torch::autograd::SavedVariable> saved_variables;
+
+  std::unordered_set<at::TensorImpl*> non_differentiable;
+  std::unordered_set<at::TensorImpl*> dirty_inputs;
```
Fields in the struct should have a trailing underscore. This is especially important because these count as the public API (though, in your next patch, you make them private, so it is less of a big deal.)

Commit:
Update on "Support custom autograd functions in C++"


Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of CFunction, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. For example: 

```
class MyFunction : public CFunction<MyFunction> {
  static variable_list forward(AutogradContext *ctx, int n, Variable var);
  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

This is a first draft of what the API will look like. 
Planned updates to this stack: 
1. ~~Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`.~~ #23184 
2. ~~Support for non variable inputs to `forward` and to save in context~~
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

Commit:
Update on "Support custom autograd functions in C++"


Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. For example: 

```
class MyFunction : public Function<MyFunction> {
  static variable_list forward(AutogradContext *ctx, int n, Variable var);
  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

This is a first draft of what the API will look like. 
Planned updates to this stack: 
1. ~~Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`.~~ #23184 
2. ~~Support for non variable inputs to `forward` and to save in context~~
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

Commit:
Update on "Support custom autograd functions in C++"


Summary:
Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. For example: 

```
class MyFunction : public Function<MyFunction> {
  static variable_list forward(AutogradContext *ctx, int n, Variable var);
  static variable_list backward(AutogradContext *ctx, variable_list grad_output);
};
```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

This is a first draft of what the API will look like. 
Planned updates to this stack: 
1. ~~Add `save_for_backward`, `mark_dirty` and `mark_non_differentiable` methods to `AutogradContext`.~~ #23184 
2. ~~Support for non variable inputs to `forward` and to save in context~~
3. More detailed documentation 
4. C++ Tests

Test Plan:
Will add tests based on the python tests

malvika2147 commented:
squashed the entire stack into a new PR
